File: 2022/typescript/src/day02.ts
function run(input: string): void {
    const res1 = part1(input);
    console.log('Day 02, part 1:', res1);
    const res2 = part2(input);
    console.log('Day 02, part 2:', res2);
}

function part1(input: string): number {
    const lines = input.split('\n');
    const rounds = lines.map((line) => line.split(' '));
    const scores = rounds.map((round) => score(round));
    const sumB = scores.reduce((acc, score) => acc + score[1], 0);
    return sumB;
}

function score(round: string[]): number[] {
    const [a, b] = round;
    const aScore = charToNumber(a);
    const bScore = charToNumber(b);

    let res: number[] = [];
    if (aScore - bScore === 0) res = [aScore + 3, bScore + 3];

    // if (aScore === 1 && bScore === 2) res = [aScore, bScore + 6]; // -1
    // if (aScore === 1 && bScore === 3) res = [aScore + 6, bScore]; // -2
    // if (aScore === 2 && bScore === 3) res = [aScore, bScore + 6]; // -1
    // if (aScore === 2 && bScore === 1) res = [aScore + 6, bScore]; // 1
    // if (aScore === 3 && bScore === 1) res = [aScore, bScore + 6]; // 2
    // if (aScore === 3 && bScore === 2) res = [aScore + 6, bScore]; // 1

    const diff = aScore - bScore;
    if (diff === -2 || diff === 1) res = [aScore + 6, bScore];
    if (diff === -1 || diff === 2) res = [aScore, bScore + 6];

    return res;
}

function charToNumber(char: string): number {
    if (char === 'A' || char === 'X') return 1;
    if (char === 'B' || char === 'Y') return 2;
    if (char === 'C' || char === 'Z') return 3;
    return 0;
}

function part2(input: string): number {
    const lines = input.split('\n');
    const rounds = lines.map((line) => line.split(' '));
    const newRounds = rounds.map((round) => strategy_guide(round));
    const scores = newRounds.map((round) => score(round));
    const sumB = scores.reduce((acc, score) => acc + score[1], 0);
    return sumB;
}

function strategy_guide(round: string[]): string[] {
    const [a, b] = round;
    const strategyB = strategyToChar(a, b);
    return [a, strategyB];
}

function strategyToChar(a: string, b: string): string {
    if (b === 'X') return shiftChar(a, - 1);
    if (b === 'Y') return a;
    if (b === 'Z') return shiftChar(a, + 1);
    return '';
}

function shiftChar(char: string, shift: number): string {
    const num = charToNumber(char);
    if (num === 2) return numberToChar(num + shift);
    if (num === 1 && shift === -1) return numberToChar(3);
    if (num === 3 && shift === +1) return numberToChar(1);
    return numberToChar(num + shift);
}

function numberToChar(num: number): string {
    if (num === 1) return 'A';
    if (num === 2) return 'B';
    if (num === 3) return 'C';
    return '';
}
export default { run, part1, part2 };

File: 2022/fsharp/days/Day02.fs
module Day02

let charToNumber (char: string) =
    if char = "A" || char = "X" then 1
    elif char = "B" || char = "Y" then 2
    elif char = "C" || char = "Z" then 3
    else 0

let score (round: string[]) =
    let chars = round |> Seq.toList
    let themScore = charToNumber chars.[0]
    let meScore = charToNumber chars.[1]
    let diff = themScore - meScore

    if diff = -2 || diff = 1 then [ themScore + 6; meScore ]
    elif diff = -1 || diff = 2 then [ themScore; meScore + 6 ]
    else [ themScore + 3; meScore + 3 ]

let part1 (input: string) =
    let lines = input.Split("\n")
    let rounds = lines |> Array.map (fun line -> line.Split(" "))
    let scores = rounds |> Array.map score
    let sumB = scores |> Array.sumBy (fun score -> score.[1])
    sumB

let numberToMeChar (num: int) =
    match num with
    | 1 -> "A"
    | 2 -> "B"
    | 3 -> "C"
    | _ -> ""

let shiftChar (char: string) (shift: int) =
    let num = charToNumber char

    match (num, shift) with
    | 2, _ -> numberToMeChar (num + shift)
    | 1, -1 -> numberToMeChar 3
    | 3, 1 -> numberToMeChar 1
    | _, _ -> numberToMeChar (num + shift)

let strategyToChar (them: string) (me: string) =
    match me with
    | "X" -> shiftChar them -1
    | "Y" -> them
    | "Z" -> shiftChar them +1
    | _ -> ""

let strategy_guide (round: string[]) =
    let chars = round |> Seq.toList
    let them = chars.[0]
    let me = chars.[1]
    let strategyB = strategyToChar them me
    [ them; strategyB ] |> Array.ofList

let part2 (input: string) =
    let lines = input.Split("\n")
    let rounds = lines |> Array.map (fun line -> line.Split(" "))
    let newRounds = rounds |> Array.map strategy_guide
    let scores = newRounds |> Array.map score
    let sumB = scores |> Array.sumBy (fun score -> score.[1])
    sumB

let run (input: string) =
    printfn "Day 02, part1: %A" (part1 input)
    printfn "Day 02, part2: %A" (part2 input)
